### ISSUE TITLE
Legendre, Jacobi and Kronecker symbols

### ISSUE THREAD

**Issue by thofma** [Nemocas/Nemo.jl#998]
First of all, we don't have the Legendre and Kronecker symbols, because flint does not have it I guess.

Now the `fmpz_jacobi` docstring says that for `(a, b)` the number  `b` must be a prime. But this would make it the Legendre symbol? On the other hand, `fmpz_jacobi` is calling `mpz_jacobi`, which really is the Jacobi symbol (see https://gmplib.org/manual/Number-Theoretic-Functions) and thus is restricted to `b` odd, but not necessarily prime. There is some discrepancy between the docstring and the implementation. Not sure if we can just change the docstring. For this one would have to check what `n_jacobi` is doing.

What should we do about this? Should I move this issue to the flint repository?

**Comment by wbhart** [comment]
n_jacobi requires b odd and positive and coprime to a for performance reasons. That's why fmpz_jacobi is currently limited to b prime and presumably a reduced mod b.

I agree that fmpz_jacobi should be reimplemented to actually give the full jacobi symbol. That'll cost an additional gcd if we want to compute it using n_jacobi I guess.

I'd open a Flint ticket for this.

**Comment by thofma** [comment]
Thanks for looking into this. The docstring of `n_jacobi` does not mention prime, just `gcd(x, y) == 1`. Is there a mismatch with the implementation or did I misunderstand something?

**Comment by wbhart** [comment]
No, the restriction to prime is deliberate, but unnecessary. It's just to make the description simple. Otherwise it is complicated to give a full description of the limitations.

**Comment by thofma** [comment]
Any idea how we could get a working `jacobi_symbol` into Nemo without waiting for the next flint release? I have a working hack in Hecke, but I would rather use the flint one.

**Comment by wbhart** [comment]
Not really. I mean, Nemo could take the gcd and reduce the numerator and make sure the denominator is positive. But it pretty much needs to be fixed in Flint if you want something better.

**Comment by thofma** [comment]
I guess this would be good enough for now. I will probably do this and make a PR if there is no objection.

**Comment by wbhart** [comment]
There is now https://github.com/wbhart/flint2/issues/895

**Comment by wbhart** [comment]
I'm happy to tag a dev version on flint2/trunk once @tthsqe12 is done with https://github.com/wbhart/flint2/pull/893

This is not going to save us from issuing lots of releases though. If you want that let me know and I'll change the milestone to 2.7.2-dev and make sure the new fmpz_jacobi goes into that.

**Comment by tthsqe12** [comment]
What is the goal here? To have working fmpz_{legendre|jacobi|kronecker} before 2.8?

**Comment by wbhart** [comment]
No, I think just get fmpz_jacobi to work for odd, positive denominator I think. I'm not sure whether to bother with negative denominator or not. Technically Jacobi didn't define that, Kronecker did.

**Comment by wbhart** [comment]
It's also not clear what the best strategy is: take gcd first to reduce the fraction or just go straight for quadratic reciprocity and bomb out if a common factor is detected. The current n_jacobi is insanely fast, so it's not clear at all which would be best. I might have to try both.

**Comment by thofma** [comment]
Yeah, not sure what the best strategy is.

Regarding what the goal is, at least in Nemo we should be able to do 
```
jacobi_symbol(a::fmpz, b::fmpz)
```
with the requirement `b >= 3`. We can derive from this easily `kronecker_symbol` or `legendre_symbol` on the Nemo side. We can also add a `reduced = false` argument if we think there should be a faster way for a user to call the function with reduced input.

I guess the flint version we would tag would be binary compatible with 2.7.0.? In this case there is no need to rebuild all the other  package I think.

**Comment by wbhart** [comment]
I think it would be compatible, yes. The main changes are fmpz_mod_mpoly, which we don't have in Nemo anyway.

**Comment by tthsqe12** [comment]
I wouldn't use the dev version of flint. I would just hack together a working legendre/jacobi/kronecker symbol in nemo, which will probably be good enough.

**Comment by tthsqe12** [comment]
@thofma have you experienced any wrong answers from fmpz_jacobi? The code for n_jacobi looks correct and the test code looks good too. The only thing I have to complain about is n_jacobi being 2x slower than necessary.
TLDR: as far as I can see fmpz_jacobi(n, m) works for any n and any positive odd m: the only thing wrong is the documentation.

**Comment by thofma** [comment]
This was exactly my original question. I will try tomorrow. By any n, do you mean any n or n reduced modulo m?

**Comment by tthsqe12** [comment]
Any n = _any_ n

**Comment by tthsqe12** [comment]
🤦‍♂️🤦‍♂️🤦‍♂️
The code for n_jacobi(x,y) is correct in all cases where jacobi is defined. But, its documentation is unnecessary restrictive, which caused fmpz_jacobi to be unnecessarily restrictive. So, fmpz_jacobi is not correct in some corner cases, where silly code was added based on the unnecessary assumptions.

**Comment by wbhart** [comment]
All that remains is to implement Legendre and Kronecker symbols.

**Comment by wbhart** [comment]
In fact, checking whether the denominator is prime will take longer than any time saved in an implementation of the Legendre symbol, so we really only need the Kronecker symbol.
